File: ios/Sources/UnifiedTrackingPlugin/ErrorTracking/SentryProvider.swift
import Foundation
import Sentry

/// Sentry (sentry-cocoa) bridge for the unified-tracking plugin.
///
/// Reads `dsn`, `environment`, and `release` from the provider config and starts
/// the Sentry SDK. If no DSN is supplied the SDK is not started and the provider
/// stays uninitialized (so ProviderManager will skip it).
class SentryProvider: ErrorTrackingProvider {
    let name = "sentry"
    private(set) var isEnabled: Bool
    private(set) var isInitialized: Bool

    private let config: [String: Any]
    private var debugMode = false
    /// Tracks the consent state. Sentry has no fine-grained per-event consent gate,
    /// so we close the SDK when consent is revoked (see `setConsent`).
    private var consentGranted = true

    init(config: [String: Any]) {
        self.config = config
        self.isEnabled = config["enabled"] as? Bool ?? true
        self.isInitialized = false

        guard let dsn = config["dsn"] as? String, !dsn.isEmpty else {
            // No DSN -> cannot start Sentry. Leave uninitialized.
            print("[SentryProvider] No DSN provided; Sentry not started.")
            return
        }

        let environment = config["environment"] as? String
        let release = config["release"] as? String

        // NOTE(unverified): build with Xcode — confirm the `SentrySDK.start { options in ... }`
        // closure-config API and the `options.dsn` / `options.environment` /
        // `options.releaseName` / `options.debug` property names in sentry-cocoa ~8.42.
        SentrySDK.start { options in
            options.dsn = dsn
            if let environment = environment {
                options.environment = environment
            }
            if let release = release {
                options.releaseName = release
            }
            options.debug = false
        }

        self.isInitialized = true
    }

    func logError(_ error: Error, context: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted else { return }
        // NOTE(unverified): build with Xcode — confirm `SentrySDK.capture(error:block:)`
        // with a `Scope` closure and `scope.setExtra(value:key:)` argument labels in
        // sentry-cocoa ~8.42.
        SentrySDK.capture(error: error) { scope in
            for (key, value) in context {
                scope.setExtra(value: value, key: key)
            }
        }
    }

    func setUserContext(_ userId: String, properties: [String: Any]) {
        guard isEnabled && isInitialized else { return }
        // NOTE(unverified): build with Xcode — confirm `Sentry.User(userId:)` initializer
        // and the `.data` property name in sentry-cocoa ~8.42 (the SDK's user model is
        // `Sentry.User`, qualified here to avoid any name collision).
        let user = Sentry.User(userId: userId)
        if !properties.isEmpty {
            user.data = properties
        }
        SentrySDK.setUser(user)
    }

    func setConsent(_ granted: Bool) {
        consentGranted = granted
        // NOTE: sentry-cocoa has no per-event opt-out switch. When consent is revoked
        // we close the SDK (stops all delivery). It cannot be re-opened without
        // re-running `init`/`SentrySDK.start`, so re-granting consent at runtime would
        // require re-initializing the provider — a documented limitation.
        if !granted {
            SentrySDK.close()
        }
    }

    func reset() {
        SentrySDK.setUser(nil)
    }

    func setDebugMode(_ enabled: Bool) {
        // Sentry debug logging is fixed at SDK start (`options.debug`). We retain the
        // flag here for parity; flipping it at runtime is not supported by the SDK.
        debugMode = enabled
    }
}

File: android/src/main/java/com/aoneahsan/unifiedtracking/providers/errortracking/SentryProvider.java
package com.aoneahsan.unifiedtracking.providers.errortracking;

import android.content.Context;
import android.util.Log;
import com.aoneahsan.unifiedtracking.providers.ErrorTrackingProvider;
import com.getcapacitor.JSObject;
import io.sentry.Sentry;
import io.sentry.android.core.SentryAndroid;
import io.sentry.protocol.User;
import java.util.Map;

/**
 * Sentry error-tracking provider for Android.
 *
 * Uses {@code io.sentry.android.core.SentryAndroid} for SDK initialization and the
 * static {@code io.sentry.Sentry} facade for capturing exceptions, setting user
 * context, and managing scope.
 *
 * NOTE(unverified): built against the Sentry Android SDK 7.x API. Confirm exact
 * signatures (SentryAndroid.init overloads, SentryOptions setters, User fields,
 * Sentry.captureException/configureScope) in Android Studio against the resolved
 * SDK version.
 */
public class SentryProvider implements ErrorTrackingProvider {

    private static final String TAG = "Sentry";
    private final Context context;
    private final JSObject config;
    private boolean enabled;
    private boolean initialized;
    private boolean debugMode = false;

    public SentryProvider(Context context, JSObject config) {
        this.context = context;
        this.config = config;
        this.enabled = config.getBoolean("enabled", true);
        this.initialized = false;

        try {
            String dsn = config.getString("dsn");
            if (dsn == null || dsn.trim().isEmpty()) {
                Log.e(TAG, "Sentry DSN is missing; provider will be inert.");
                return;
            }

            final String environment = config.getString("environment", null);
            final String release = config.getString("release", null);
            // tracesSampleRate is a double in the JS config; fall back to 0.0 (errors only).
            final Double tracesSampleRate = config.has("tracesSampleRate") ? config.getDouble("tracesSampleRate") : null;
            final boolean debugConfig = config.getBoolean("debug", false);

            // NOTE(unverified): SentryAndroid.init(Context, OptionsConfiguration) is the
            // documented Android entry point in SDK 7.x. Confirm the lambda receives
            // SentryAndroidOptions (a subclass of SentryOptions) in Android Studio.
            SentryAndroid.init(context, (options) -> {
                options.setDsn(dsn);
                if (environment != null && !environment.trim().isEmpty()) {
                    options.setEnvironment(environment);
                }
                if (release != null && !release.trim().isEmpty()) {
                    options.setRelease(release);
                }
                if (tracesSampleRate != null) {
                    options.setTracesSampleRate(tracesSampleRate);
                }
                options.setDebug(debugConfig);
            });

            this.initialized = true;

            if (debugMode) {
                Log.d(TAG, "Sentry initialized (environment=" + environment + ")");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Sentry", e);
            this.initialized = false;
        }
    }

    @Override
    public String getName() {
        return "sentry";
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public boolean isInitialized() {
        return initialized;
    }

    @Override
    public void logError(Exception error, Map<String, Object> context) {
        if (!enabled || !initialized || error == null) return;

        try {
            // Attach the context map as scope "extras" so it travels with the event,
            // then capture the throwable. configureScope mutates the current scope.
            if (context != null && !context.isEmpty()) {
                Sentry.configureScope((scope) -> {
                    for (Map.Entry<String, Object> entry : context.entrySet()) {
                        if (entry.getValue() == null) continue;
                        // NOTE(unverified): Scope.setExtra(String, Object) exists in SDK 7.x.
                        // Some keys (e.g. "level", "tags") could be promoted to first-class
                        // fields; here everything is treated uniformly as an extra.
                        scope.setExtra(entry.getKey(), String.valueOf(entry.getValue()));
                    }
                });
            }

            Sentry.captureException(error);

            if (debugMode) {
                Log.d(TAG, "Captured exception: " + error.getMessage());
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to capture exception", e);
        }
    }

    @Override
    public void setUserContext(String userId, Map<String, Object> properties) {
        if (!enabled || !initialized) return;

        try {
            User user = new User();
            user.setId(userId);

            if (properties != null) {
                Object email = properties.get("email");
                if (email != null) {
                    user.setEmail(String.valueOf(email));
                }
                Object username = properties.get("username");
                if (username != null) {
                    user.setUsername(String.valueOf(username));
                }
                // NOTE(unverified): User.setData(Map<String,String>) is the documented
                // way to attach arbitrary user attributes in SDK 7.x. Confirm the method
                // name/signature; older SDKs used setOthers(...).
                java.util.Map<String, String> extra = new java.util.HashMap<>();
                for (Map.Entry<String, Object> entry : properties.entrySet()) {
                    String key = entry.getKey();
                    if ("email".equals(key) || "username".equals(key)) continue;
                    if (entry.getValue() == null) continue;
                    extra.put(key, String.valueOf(entry.getValue()));
                }
                if (!extra.isEmpty()) {
                    user.setData(extra);
                }
            }

            Sentry.setUser(user);

            if (debugMode) {
                Log.d(TAG, "Set user context: " + userId);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set user context", e);
        }
    }

    @Override
    public void setConsent(boolean granted) {
        if (!initialized) return;

        // NOTE(unverified): The Sentry Android SDK has no first-class runtime consent
        // toggle equivalent to Firebase's setAnalyticsCollectionEnabled. We gate dispatch
        // locally via the `enabled` flag (logError/setUserContext early-return when false).
        // The SDK itself stays initialized; to fully stop the SDK at runtime you would need
        // Sentry.close() and re-init, which is intentionally avoided here to keep behavior simple.
        this.enabled = granted;

        if (debugMode) {
            Log.d(TAG, "Set consent (local gate): " + granted);
        }
    }

    @Override
    public void reset() {
        if (!initialized) return;

        try {
            // Clear the associated user; subsequent events are anonymous.
            Sentry.setUser(null);

            if (debugMode) {
                Log.d(TAG, "Reset user context");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to reset", e);
        }
    }

    @Override
    public void setDebugMode(boolean enabled) {
        this.debugMode = enabled;
        if (enabled) {
            Log.d(TAG, "Debug mode enabled for Sentry provider");
        }
    }
}
